cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/NamespaceSourceControlMetadataStore.java

```diff
@@ -0,0 +1,219 @@
+    Optional<StructuredRow> row = table.read(primaryKey);
+
+    return row.map(nonNullRow -> {
+      String specificationHash = nonNullRow.getString(
+          StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD);
+      String commitId = nonNullRow.getString(
+          StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD);
+      Long lastSynced = nonNullRow.getLong(
+          StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD);
+      lastSynced = lastSynced == 0L ? null : lastSynced;
+       Boolean isSynced = nonNullRow.getBoolean(
+          StoreDefinition.NamespaceSourceControlMetadataStore.IS_SYNCED_FIELD);
+      if (specificationHash == null && commitId == null && lastSynced == null) {
+        return null;
+      }
+      return new SourceControlMeta(specificationHash, commitId, Instant.ofEpochMilli(lastSynced),
+          isSynced);
+    }).orElse(null);
+  }
+
+  /**
+   * Sets the source control metadata for the specified application ID in the namespace.  Source
+   * control metadata will be null when the application is deployed in the namespace. It will be
+   * non-null when the application is pulled from the remote repository and deployed in the
+   * namespace.
+   *
+   * @param appRef            {@link ApplicationReference} for which the source control metadata is
+   *                          being set.
+   * @param sourceControlMeta The {@link SourceControlMeta} to be set. Can be {@code null} if
+   *                          application is just deployed.
+   * @throws IOException If failed to write the data.
+   */
+  public void write(ApplicationReference appRef,
+      @Nullable SourceControlMeta sourceControlMeta)
+      throws IOException {
+    // In the Namespace Pipelines page, the sync status (SYNCED or UNSYNCED)
+    // and last modified of all the applications deployed in the namespace needs to be shown.
+    // If source control information is not added when the app is deployed, the data will
+    // split into two tables.
+    // JOIN operation is not currently supported yet. The filtering (eg,
+    // filter on UNSYNCED sync status) , sorting, searching , pagination becomes difficult.
+    // Instead of doing filtering, searching, sorting in memory, it will happen at
+    // database level.
+    StructuredTable scmTable = getNamespaceSourceControlMetadataTable();
+    SourceControlMeta existingSourceControlMeta = get(appRef);
+    if (sourceControlMeta != null && existingSourceControlMeta != null
+        && sourceControlMeta.getLastSyncedAt()
+        .isBefore(existingSourceControlMeta.getLastSyncedAt())) {
+      List<Field<?>> fields = new ArrayList<>();
+      fields.add(
+          Fields.longField(StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD,
+              sourceControlMeta.getLastSyncedAt().toEpochMilli()));
+      throw new InvalidFieldException(
```

Comment:
I dont think `InvalidFieldException` is meant for this usecase. Use `IllegalArgumentException` instead. Also do not need to create a field object for this case. Just write a message with the details